pkg/api/auth_middleware.go

```diff
@@ -148,11 +160,107 @@ func enhanceWithFriendlyName(user *model.User, friendlyName string) *model.User
+
+	if idTokenClaims == nil {
+		return nil, nil
+	}
+
+	if !ok {
+		logger.WithField("claims", authSession.Values[SAMLTokenClaimsSessionKey]).Error("failed decoding tokens")
+		return nil, ErrAuthenticatingRequest
+	}
+
+	logger.WithField("claims", idTokenClaims).Debug("success decoding token claims")
+
+	idKey := cookieAuthConfig.ExternalUserIDClaimName
+	externalID, ok := idTokenClaims[idKey].(string)
+	if !ok {
+		logger.WithField(idKey, idTokenClaims[idKey]).Error("failed type assertion for sub claim")
+		return nil, ErrAuthenticatingRequest
+	}
+
+	log := logger.WithField("external_id", externalID)
+
+	for claimName, expectedValue := range cookieAuthConfig.ValidateIDTokenClaims {
+		actualValue, ok := idTokenClaims[claimName]
+		if !ok || actualValue != expectedValue {
+			log.WithFields(logging.Fields{
+				"claim_name":     claimName,
+				"actual_value":   actualValue,
+				"expected_value": expectedValue,
+				"missing":        !ok,
+			}).Error("authentication failed on validating ID token claims")
+			return nil, ErrAuthenticatingRequest
+		}
+	}
+
+	// update user
+	// TODO(isan) consolidate userFromOIDC and userFromSAML below here internal db handling code
+	friendlyName := ""
+	if cookieAuthConfig.FriendlyNameClaimName != "" {
+		friendlyName, _ = idTokenClaims[cookieAuthConfig.FriendlyNameClaimName].(string)
+	}
+	log = log.WithField("friendly_name", friendlyName)
+
+	user, err := authService.GetUserByExternalID(ctx, externalID)
+	if err == nil {
+		log.Info("user found returning success ")
+		return enhanceWithFriendlyName(user, friendlyName), nil
+	}
+
+	if !errors.Is(err, auth.ErrNotFound) {
+		log.WithError(err).Error("failed while searching if user exists in database")
+		return nil, ErrAuthenticatingRequest
+	}
+	log.Info("user not found; creating them")
+
+	u := model.User{
+		CreatedAt:  time.Now().UTC(),
+		Source:     cookieAuthConfig.AuthSource,
+		Username:   externalID,
+		ExternalID: &externalID,
+	}
+
+	_, err = authService.CreateUser(ctx, &u)
+	if err != nil {
+		if !errors.Is(err, auth.ErrAlreadyExists) {
+			log.WithError(err).Error("failed to create external user in database")
+			return nil, ErrAuthenticatingRequest
+		}
+		// user already exists - get it:
+		user, err = authService.GetUserByExternalID(ctx, externalID)
+		if err != nil {
+			log.WithError(err).Error("failed to get external user from database")
+			return nil, ErrAuthenticatingRequest
+		}
+		return enhanceWithFriendlyName(user, friendlyName), nil
+	}
+	initialGroups := cookieAuthConfig.DefaultInitialGroups
+	if userInitialGroups, ok := idTokenClaims[cookieAuthConfig.InitialGroupsClaimName].(string); ok {
+		initialGroups = strings.Split(userInitialGroups, ",")
+	}
+	for _, g := range initialGroups {
+		err = authService.AddUserToGroup(ctx, u.Username, strings.TrimSpace(g))
+		if err != nil {
+			log.WithError(err).Error("Failed to add external user to group")
```

Comment:
Shouldn't we fail for this?

Reply:
I don't think so (The worst case here is that the user is not getting a group by default) compared to partial registration,   I followed the same blueprint we already have in OIDC, so I don't want to change the pattern.